Compile error with OpenCV 4

```bash
mkdir build
pushd build
cmake -D CMAKE_INSTALL_PREFIX=${HOME}/py3-caffe-refinedet -D python_version=3 -D CPU_ONLY=1 ..
make
```

### Errorlog

```bash
[ 60%] Building CXX object src/caffe/CMakeFiles/caffe.dir/layers/video_data_layer.cpp.o
/home/k-tanabe/RefineDet/src/caffe/layers/video_data_layer.cpp: In member function ‘virtual void caffe::VideoDataLayer<Dtype>::DataLayerSetUp(const std::vector<caffe::Blob<Dtype>*>&, const std::vector<caffe::Blob<Dtype>*>&)’:
/home/k-tanabe/RefineDet/src/caffe/layers/video_data_layer.cpp:55:30: error: ‘CV_CAP_PROP_FRAME_COUNT’ was not declared in this scope
   55 |     total_frames_ = cap_.get(CV_CAP_PROP_FRAME_COUNT);
      |                              ^~~~~~~~~~~~~~~~~~~~~~~
/home/k-tanabe/RefineDet/src/caffe/layers/video_data_layer.cpp:60:14: error: ‘CV_CAP_PROP_POS_FRAMES’ was not declared in this scope
   60 |     cap_.set(CV_CAP_PROP_POS_FRAMES, 0);
      |              ^~~~~~~~~~~~~~~~~~~~~~
make[2]: *** [src/caffe/CMakeFiles/caffe.dir/build.make:1168: src/caffe/CMakeFiles/caffe.dir/layers/video_data_layer.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:361: src/caffe/CMakeFiles/caffe.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
```

--- a/src/caffe/layers/video_data_layer.cpp
+++ b/src/caffe/layers/video_data_layer.cpp
@@ -13,6 +13,11 @@
 #include "caffe/layers/video_data_layer.hpp"
 #include "caffe/util/benchmark.hpp"
 
+#if (CV_MAJOR_VERSION >= 4)
+#define CV_CAP_PROP_FRAME_COUNT cv::CAP_PROP_FRAME_COUNT
+#define CV_CAP_PROP_POS_FRAMES cv::CAP_PROP_POS_FRAMES
+#endif
+
 namespace caffe {
 
 template <typename Dtype>

--- a/src/caffe/layers/window_data_layer.cpp
+++ b/src/caffe/layers/window_data_layer.cpp
@@ -290,7 +290,7 @@ void WindowDataLayer<Dtype>::load_batch(Batch<Dtype>* batch) {
           image_database_cache_[window[WindowDataLayer<Dtype>::IMAGE_INDEX]];
         cv_img = DecodeDatumToCVMat(image_cached.second, true);
       } else {
-        cv_img = cv::imread(image.first, CV_LOAD_IMAGE_COLOR);
+        cv_img = cv::imread(image.first, cv::IMREAD_COLOR);
         if (!cv_img.data) {
           LOG(ERROR) << "Could not open or find file " << image.first;
           return;

--- a/src/caffe/test/test_io.cpp
+++ b/src/caffe/test/test_io.cpp
@@ -20,8 +20,8 @@ class IOTest : public ::testing::Test {};
 bool ReadImageToDatumReference(const string& filename, const int label,
     const int height, const int width, const bool is_color, Datum* datum) {
   cv::Mat cv_img;
-  int cv_read_flag = (is_color ? CV_LOAD_IMAGE_COLOR :
-    CV_LOAD_IMAGE_GRAYSCALE);
+  int cv_read_flag = (is_color ? cv::IMREAD_COLOR :
+    cv::IMREAD_GRAYSCALE);
 
   cv::Mat cv_img_origin = cv::imread(filename, cv_read_flag);
   if (!cv_img_origin.data) {

--- a/src/caffe/util/bbox_util.cpp
+++ b/src/caffe/util/bbox_util.cpp
@@ -12,6 +12,11 @@
 
 #include "caffe/util/bbox_util.hpp"
 
+#if (CV_MAJOR_VERSION >= 4)
+#define CV_FILLED cv::FILLED
+#define CV_FOURCC cv::VideoWriter::fourcc
+#endif
+
 namespace caffe {
 
 bool SortBBoxAscend(const NormalizedBBox& bbox1, const NormalizedBBox& bbox2) {

--- a/src/caffe/util/im_transforms.cpp
+++ b/src/caffe/util/im_transforms.cpp
@@ -243,12 +243,12 @@ void CenterObjectAndFillBg(const cv::Mat& in_img, const bool fill_bg,
   cv::Mat mask, crop_mask;
   if (in_img.channels() > 1) {
     cv::Mat in_img_gray;
-    cv::cvtColor(in_img, in_img_gray, CV_BGR2GRAY);
+    cv::cvtColor(in_img, in_img_gray, cv::COLOR_BGR2GRAY);
     cv::threshold(in_img_gray, mask, 0, 255,
-                  CV_THRESH_BINARY_INV | CV_THRESH_OTSU);
+                  cv::THRESH_BINARY_INV | cv::THRESH_OTSU);
   } else {
     cv::threshold(in_img, mask, 0, 255,
-                  CV_THRESH_BINARY_INV | CV_THRESH_OTSU);
+                  cv::THRESH_BINARY_INV | cv::THRESH_OTSU);
   }
   cv::Rect crop_rect = CropMask(mask, mask.at<uchar>(0, 0), 2);
 
@@ -426,8 +426,8 @@ cv::Mat ApplyNoise(const cv::Mat& in_img, const NoiseParameter& param) {
 
   if (param.decolorize()) {
     cv::Mat grayscale_img;
-    cv::cvtColor(in_img, grayscale_img, CV_BGR2GRAY);
-    cv::cvtColor(grayscale_img, out_img,  CV_GRAY2BGR);
+    cv::cvtColor(in_img, grayscale_img, cv::COLOR_BGR2GRAY);
+    cv::cvtColor(grayscale_img, out_img,  cv::COLOR_GRAY2BGR);
   } else {
     out_img = in_img;
   }
@@ -439,7 +439,7 @@ cv::Mat ApplyNoise(const cv::Mat& in_img, const NoiseParameter& param) {
   if (param.hist_eq()) {
     if (out_img.channels() > 1) {
       cv::Mat ycrcb_image;
-      cv::cvtColor(out_img, ycrcb_image, CV_BGR2YCrCb);
+      cv::cvtColor(out_img, ycrcb_image, cv::COLOR_BGR2YCrCb);
       // Extract the L channel
       vector<cv::Mat> ycrcb_planes(3);
       cv::split(ycrcb_image, ycrcb_planes);
@@ -449,7 +449,7 @@ cv::Mat ApplyNoise(const cv::Mat& in_img, const NoiseParameter& param) {
       ycrcb_planes[0] = dst;
       cv::merge(ycrcb_planes, ycrcb_image);
       // convert back to RGB
-      cv::cvtColor(ycrcb_image, out_img, CV_YCrCb2BGR);
+      cv::cvtColor(ycrcb_image, out_img, cv::COLOR_YCrCb2BGR);
     } else {
       cv::Mat temp_img;
       cv::equalizeHist(out_img, temp_img);
@@ -462,7 +462,7 @@ cv::Mat ApplyNoise(const cv::Mat& in_img, const NoiseParameter& param) {
     clahe->setClipLimit(4);
     if (out_img.channels() > 1) {
       cv::Mat ycrcb_image;
-      cv::cvtColor(out_img, ycrcb_image, CV_BGR2YCrCb);
+      cv::cvtColor(out_img, ycrcb_image, cv::COLOR_BGR2YCrCb);
       // Extract the L channel
       vector<cv::Mat> ycrcb_planes(3);
       cv::split(ycrcb_image, ycrcb_planes);
@@ -472,7 +472,7 @@ cv::Mat ApplyNoise(const cv::Mat& in_img, const NoiseParameter& param) {
       ycrcb_planes[0] = dst;
       cv::merge(ycrcb_planes, ycrcb_image);
       // convert back to RGB
-      cv::cvtColor(ycrcb_image, out_img, CV_YCrCb2BGR);
+      cv::cvtColor(ycrcb_image, out_img, cv::COLOR_YCrCb2BGR);
     } else {
       cv::Ptr<cv::CLAHE> clahe = cv::createCLAHE();
       clahe->setClipLimit(4);
@@ -485,10 +485,10 @@ cv::Mat ApplyNoise(const cv::Mat& in_img, const NoiseParameter& param) {
   if (param.jpeg() > 0) {
     vector<uchar> buf;
     vector<int> params;
-    params.push_back(CV_IMWRITE_JPEG_QUALITY);
+    params.push_back(cv::IMWRITE_JPEG_QUALITY);
     params.push_back(param.jpeg());
     cv::imencode(".jpg", out_img, buf, params);
-    out_img = cv::imdecode(buf, CV_LOAD_IMAGE_COLOR);
+    out_img = cv::imdecode(buf, cv::IMREAD_COLOR);
   }
 
   if (param.erode()) {
@@ -536,14 +536,14 @@ cv::Mat ApplyNoise(const cv::Mat& in_img, const NoiseParameter& param) {
 
   if (param.convert_to_hsv()) {
     cv::Mat hsv_image;
-    cv::cvtColor(out_img, hsv_image, CV_BGR2HSV);
+    cv::cvtColor(out_img, hsv_image, cv::COLOR_BGR2HSV);
     out_img = hsv_image;
   }
   if (param.convert_to_lab()) {
     cv::Mat lab_image;
     out_img.convertTo(lab_image, CV_32F);
     lab_image *= 1.0 / 255;
-    cv::cvtColor(lab_image, out_img, CV_BGR2Lab);
+    cv::cvtColor(lab_image, out_img, cv::COLOR_BGR2Lab);
   }
   return  out_img;
 }
@@ -614,7 +614,7 @@ void AdjustSaturation(const cv::Mat& in_img, const float delta,
                       cv::Mat* out_img) {
   if (fabs(delta - 1.f) != 1e-3) {
     // Convert to HSV colorspae.
-    cv::cvtColor(in_img, *out_img, CV_BGR2HSV);
+    cv::cvtColor(in_img, *out_img, cv::COLOR_BGR2HSV);
 
     // Split the image to 3 channels.
     vector<cv::Mat> channels;
@@ -625,7 +625,7 @@ void AdjustSaturation(const cv::Mat& in_img, const float delta,
     cv::merge(channels, *out_img);
 
     // Back to BGR colorspace.
-    cvtColor(*out_img, *out_img, CV_HSV2BGR);
+    cvtColor(*out_img, *out_img, cv::COLOR_HSV2BGR);
   } else {
     *out_img = in_img;
   }
@@ -648,7 +648,7 @@ void RandomHue(const cv::Mat& in_img, cv::Mat* out_img,
 void AdjustHue(const cv::Mat& in_img, const float delta, cv::Mat* out_img) {
   if (fabs(delta) > 0) {
     // Convert to HSV colorspae.
-    cv::cvtColor(in_img, *out_img, CV_BGR2HSV);
+    cv::cvtColor(in_img, *out_img, cv::COLOR_BGR2HSV);
 
     // Split the image to 3 channels.
     vector<cv::Mat> channels;
@@ -659,7 +659,7 @@ void AdjustHue(const cv::Mat& in_img, const float delta, cv::Mat* out_img) {
     cv::merge(channels, *out_img);
 
     // Back to BGR colorspace.
-    cvtColor(*out_img, *out_img, CV_HSV2BGR);
+    cvtColor(*out_img, *out_img, cv::COLOR_HSV2BGR);
   } else {
     *out_img = in_img;
   }

--- a/src/caffe/util/io.cpp
+++ b/src/caffe/util/io.cpp
@@ -83,8 +83,8 @@ cv::Mat ReadImageToCVMat(const string& filename, const int height,
     const int width, const int min_dim, const int max_dim,
     const bool is_color) {
   cv::Mat cv_img;
-  int cv_read_flag = (is_color ? CV_LOAD_IMAGE_COLOR :
-    CV_LOAD_IMAGE_GRAYSCALE);
+  int cv_read_flag = (is_color ? cv::IMREAD_COLOR :
+    cv::IMREAD_GRAYSCALE);
   cv::Mat cv_img_origin = cv::imread(filename, cv_read_flag);
   if (!cv_img_origin.data) {
     LOG(ERROR) << "Could not open or find file " << filename;
@@ -663,8 +663,8 @@ cv::Mat DecodeDatumToCVMat(const Datum& datum, bool is_color) {
   CHECK(datum.encoded()) << "Datum not encoded";
   const string& data = datum.data();
   std::vector<char> vec_data(data.c_str(), data.c_str() + data.size());
-  int cv_read_flag = (is_color ? CV_LOAD_IMAGE_COLOR :
-    CV_LOAD_IMAGE_GRAYSCALE);
+  int cv_read_flag = (is_color ? cv::IMREAD_COLOR :
+    cv::IMREAD_GRAYSCALE);
   cv_img = cv::imdecode(vec_data, cv_read_flag);
   if (!cv_img.data) {
     LOG(ERROR) << "Could not decode datum ";

